pkg/controlplane/apiserver/apis.go

```diff
@@ -31,6 +44,46 @@ type RESTStorageProvider interface {
 	NewRESTStorage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (genericapiserver.APIGroupInfo, error)
 }
 
+// NewCoreGenericConfig returns a new core rest generic config.
+func (c *CompletedConfig) NewCoreGenericConfig() *corerest.GenericConfig {
+	return &corerest.GenericConfig{
+		StorageFactory:              c.Extra.StorageFactory,
+		EventTTL:                    c.Extra.EventTTL,
+		LoopbackClientConfig:        c.Generic.LoopbackClientConfig,
+		ServiceAccountIssuer:        c.Extra.ServiceAccountIssuer,
+		ExtendExpiration:            c.Extra.ExtendExpiration,
+		ServiceAccountMaxExpiration: c.Extra.ServiceAccountMaxExpiration,
+		APIAudiences:                c.Generic.Authentication.APIAudiences,
+		Informers:                   c.Extra.VersionedInformers,
+	}
+}
+
+// GenericStorageProviders returns a set of APIs for a generic control plane.
+// They ought to be a subset of those served by kube-apiserver.
+func (c *CompletedConfig) GenericStorageProviders(discovery discovery.DiscoveryInterface) ([]RESTStorageProvider, error) {
+	// The order here is preserved in discovery.
+	// If resources with identical names exist in more than one of these groups (e.g. "deployments.apps"" and "deployments.extensions"),
+	// the order of this list determines which group an unqualified resource name (e.g. "deployments") should prefer.
+	// This priority order is used for local discovery, but it ends up aggregated in `k8s.io/kubernetes/cmd/kube-apiserver/app/aggregator.go
+	// with specific priorities.
+	// TODO: describe the priority all the way down in the RESTStorageProviders and plumb it back through the various discovery
+	// handlers that we have.
+	return []RESTStorageProvider{
```

Comment:
Ideas welcome of how to make this helpful. This is an ordered subset of the kube rest storage provider list. We want the same order. And we want to declare the construction only once if possible. Any good pattern for that?

Reply:
These are quite static ones. Would something like controllers "Register pattern" work?
```
controlplaneapiserver.Register(Providers...)
controlplaneapiserver.GetProviders() 
```
and we can make sure the order (and in the future priority) is honored inside GetProviders?  

regards the priority, feels like main blocker is that TODO in the end . Need to think more TBH...

Reply:
maybe whats here is fine for now and we see how painful it will be to "control planes authors" and see what we can do? Don't know...

Reply:
I haven't checked the whole codepath out, but it feels like the best way to do this would be to after this split actually try to go for the TODO mentioned, to plumb through the api serving version priorities, and based of that, build this order as well.

good at least that this code is net-new and not used by k8s, after merging we can see what ends up being the best pattern for registering this. as it's "just" a slice anyways, a power-user could edit it manually, or choose not just to call it for now. I believe most users will anyways just append to this list, and not have conflicting resource names with "core generic kube".

Reply:
this ^ :) 
ship and revisti @sttts ?

Reply:
Would like to see a sketch before moving forward. This part really matters as everybody will have to select the storage providers, i.e. construct them and put them into an order. This should be as automatic as possible such that we can easily change anything here and nobody has to update their code.

Reply:
I imagine if we extend 
```
type RESTStorageProvider interface {
	GroupName() string
	NewRESTStorage(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter) (genericapiserver.APIGroupInfo, error)
	Priority() int
}
```
it will be a breaking change as everybody would have to add it? New wrapper `PrioriorityRestStorageProvider` extending old + priority? How would you imagine this TODO? Would this not be separate enhancement for ordered storage providers with defined levels and orders?

Reply:
I think where im leaning towards is:
how do you imagine this TODO being done as it might impact what we do here? 
Would each provider need to extend to new Interface? Or add new one?